### PR TITLE
create filter for setting custom span name for distributed tracing

### PR DESCRIPTION
**Pull request by aryszka**
Creating a filter to set the name of the proxy span for distributed tracing.

- [x] Create a filter
- [x] Apply the span name in the proxy
- [x] Add a parameter to override the initial span name (currently "ingress")
- [x] docs, mark the filter experimental

### CODE REVIEW

**Comment by lmineiro** [comment]
:+1:

File: cmd/skipper/main.go [hunk 1]
```diff
@@ -129,6 +128,9 @@ const (
 	enableHopHeadersRemovalUsage         = "enables removal of Hop-Headers according to RFC-2616"
 	oauth2TokeninfoURLUsage              = "sets the default tokeninfo URL to query information about an incoming OAuth2 token in oauth2Tokeninfo filters"
 	maxAuditBodyUsage                    = "sets the max body to read to log inthe audit log body"
+
```
Comment (szuecs) [inline cmd/skipper/main.go]: nit: hmm separate it from the rest?
If you do this we should probably do it in blocks in a separate PR to make this more readable for all the things.

Comment (aryszka) [reply cmd/skipper/main.go]: definitely, i just wanted to keep the changeset in this PR small. Is it fine if I just create a separate PR for the formatting?

File: filters/tracing/spanname.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+type filter struct {
+	spanName string
+}
+
+// NewSpanName creates a filter spec for setting the name of the outgoing span. (By default "proxy".)
+//
+// 	tracingSpanName("example-operation")
+//
+// WARNING: this filter is experimental, and the name and the arguments can change until marked as stable.
```
Comment (szuecs) [inline filters/tracing/spanname.go]: Do we have a policy documented somewhere how to mark things as experimental or deprecated?
I like the idea, maybe it's enough like this and we should create a separate issue to document how to do this with introducing new features and how to deal with old stuff, which I personally mark as DEPRECATED in our code base.

File: proxy/proxy.go [hunk 3]
```diff
@@ -482,22 +488,28 @@ func WithParams(p Params) *Proxy {
 		defaultHTTPStatus = p.DefaultHTTPStatus
 	}
 
+	openTracingInitialSpan := p.OpenTracingInitialSpan
+	if openTracingInitialSpan == "" {
```
Comment (szuecs) [inline proxy/proxy.go]: This can not happen as we have a default from CLI set, or is this helping the skipper use as library?

File: skipper.go [hunk 4]
```diff
@@ -792,12 +796,17 @@ func Run(o Options) error {
...
 		proxyParams.OpenTracer = tracer
+		proxyParams.OpenTracingInitialSpan = o.OpenTracingInitialSpan
 	} else {
 		// always have a tracer available, so filter authors can rely on the
 		// existence of a tracer
 		proxyParams.OpenTracer, _ = tracing.LoadTracingPlugin(o.PluginDirs, []string{"noop"})
 	}
 
+	if proxyParams.OpenTracingInitialSpan != "" {
+		proxyParams.OpenTracingInitialSpan = o.OpenTracingInitialSpan
```
Comment (szuecs) [inline skipper.go]: this should catch the skipper use as library, not?

File: filters/tracing/spanname.go (hunk 2 above)
Comment (aryszka) [reply filters/tracing/spanname.go]: we don't have currently such a policy, and yes it would be useful. I like the godoc for this, because that's something that we can assume someone always reads. How about using all-caps labels: EXPERIMENTAL, DEPRECATED, and the reason, in its own paragraph of the related godoc? e.g:

```
// NewFoo creates and initializes a new foo object. Use the Close() method to release
// resources, once it's not required anymore.
//
// EXPERIMENTAL: this feature is in an experimental state, and can be removed or
// changed in the near future.
//
func NewFoo() foo { return 42 }
```

File: proxy/proxy.go (hunk 3 above)
Comment (aryszka) [reply proxy/proxy.go]: yes, it's for the library use case, potentially for testing, too. Not particularly useful, but it's minimal effort and code.

**Comment by szuecs** [comment]
I added one nit and a couple of questions. Code is fine, only one generic question:
Do you think to have "ingress" and "proxy" span not to be too generic?
Maybe it would be better to use "skipper-ingress" or "skipper" and "skipper-proxy-call" .
Thoughts?

File: skipper.go (hunk 4 above)
Comment (aryszka) [reply skipper.go]: well, not if someone uses the proxy package as a library. In theory, it is possible.

**Comment by aryszka** [comment]
I like the terms "ingress" and "proxy", because there's a plethora of use cases and I couldn't find a better name for these operations. They are good defaults. True that they are generic, but the potential use cases of skipper are also a lot. I prefer not to put "skipper-" in the span name, because I think the span names should be about the operation, not the entities. We should try to go with the current way, and collect some experience.

The reason for this PR is that current use case raised the need to customize the name "ingress", because for one we wanted to avoid confusion with Kubernetes ingress. And we also want to give route specific operation names for the proxy spans based on the feature. At least to experiment, how it works out.

**Comment by aryszka** [comment]
see my question regarding the EXPERIMENTAL and DEPRECATED labels ;)

File: cmd/skipper/main.go (hunk 1 above)
Comment (szuecs) [reply cmd/skipper/main.go]: yes

**Comment by szuecs** [comment]
:+1:

File: filters/tracing/spanname.go (hunk 2 above)
Comment (szuecs) [reply filters/tracing/spanname.go]: I like EXPERIMENTAL (better than WARNING) and DEPRECATED as godoc

**Comment by szuecs** [comment]
@aryszka then lets change the godoc from WARNING to EXPERIMENTAL

**Comment by lmineiro** [comment]
👍🏻

**Comment by vigneshshanmugam** [comment]
👍

**Comment by lmineiro** [comment]
:+1:

**Comment by aryszka** [comment]
:+1:
